### PR TITLE
Ports a bastardized version of the supermatter cascade from /vg/

### PR DESCRIPTION
**Pull request by KingTerrytheTerrible**
## About The Pull Request

Simple but powerful round ender, it's very slow to finish so only antags should be able to start it.
## Why It's Good For The Game

More variety in SM delaminations are always good; and this time it gives engineers a good incentive to actually fix the engine aside "oh no, engineering is gone." If they don't do it, then the entire station (and universe) is fucked. Also, it's cool
## Changelog
:cl: Various /vg/station coders, KingTerryTheTerrible
add: The Supermatter Cascade
/:cl:

### CODE REVIEW

File: code/modules/power/supermatter/supermatter_cascade_escape.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		do_teleport(A, get_turf(exit), 0, null, null, null, null, TRUE)
+		var/mob/M = A
+		to_chat(M, "<span class='big blue'>You made it. You feel as though it's all going to be OK, in the end.</span>")
+		return
+	qdel(A)
+	//if (iscarbon(A))
+	
+/obj/singularity/cascade/exit/process()
+	eat()
+	return
```
Comment (ghost) [inline code/modules/power/supermatter/supermatter_cascade_escape.dm]: redundant

File: code/modules/power/supermatter/supermatter.dm [hunk 2]
```diff
@@ -279,6 +281,9 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 	explode()
 
 /obj/machinery/power/supermatter_crystal/proc/explode()
+	if(cascade_allowed && power > CASCADE_POWER_PENALTY_THRESHOLD) //hippie start
```
Comment (ghost) [inline code/modules/power/supermatter/supermatter.dm]: the comment

File: code/modules/power/supermatter/bluespace_sea.dm [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	return ..()
+
+//Stolen, I mean, ported from VGstation
+/turf/closed/indestructible/supermatter/wall/process()
+	// Only check infrequently.
+	if(next_check>world.time)
+		return
+
+	// No more available directions? Shut down process().
+	if(avail_dirs.len==0)
```
Comment (ghost) [inline code/modules/power/supermatter/bluespace_sea.dm]: `if(!LAZYLEN(avail_dirs.len))`

File: code/modules/power/supermatter/supermatter.dm (hunk 2 above)
Comment (KingTerrytheTerrible) [reply code/modules/power/supermatter/supermatter.dm:284]: woops

**Comment by KingTerrytheTerrible** [comment]
👍 thx

**Comment by BeloneX** [comment]
> 
> 
> 👍 thx

kakascript 3 when

**Comment by KingTerrytheTerrible** [comment]
image
image
Currently, the threshold to cascade is 11000KeV which is hard to achieve, but I'm down for increasing it if found to be too low or changing it to a different type (traitor must make the crystal unstable with a special item, mining find a special crystal, etc)

**Comment by coolsurf6** [comment]
Even just a bluespace crystal. Imagine a traitor Just taking 1 tc out, chucking it in. And the SM is set on a path of destruction if it ever delams.

**Comment by KingTerrytheTerrible** [comment]
A bluespace crystal sounds good, but we'll have time to try out differnet avenues later

File: code/modules/power/supermatter/supermatter.dm [hunk 4]
```diff
@@ -279,6 +281,9 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 	explode()
 
 /obj/machinery/power/supermatter_crystal/proc/explode()
+	if(cascade_allowed && power > CASCADE_POWER_PENALTY_THRESHOLD)
+		src.cascade()
+		return //hippie end
```
Comment (qwertyquerty) [inline code/modules/power/supermatter/supermatter.dm:286]: Remove the comment here

File: code/modules/power/supermatter/supermatter_cascade_escape.dm [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	//var/datum/map_template/exit
+	var/obj/effect/landmark/beach/exit
+	icon = 'icons/obj/rift.dmi'
+	icon_state = "rift"
+	pixel_x = -128
+	pixel_y = -128
+
+/obj/singularity/cascade/exit/Initialize(mapload, starting_energy)
+	. = ..()
+	/*exit = new /datum/map_template/ruin/lavaland/biodome/beach
```
Comment (qwertyquerty) [inline code/modules/power/supermatter/supermatter_cascade_escape.dm:24]: Remove these comments

File: code/modules/power/supermatter/supermatter_cascade_escape.dm [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+/obj/singularity/cascade/exit/consume(atom/A)
+	if(istype(A, /turf/closed/indestructible/supermatter/wall))
+		return
+	if (ismob(A))
+		do_teleport(A, get_turf(exit), 0, null, null, null, null, TRUE)
+		var/mob/M = A
+		to_chat(M, "<span class='big blue'>You made it. You feel as though it's all going to be OK, in the end.</span>")
+		return
+	qdel(A)
+	//if (iscarbon(A))
```
Comment (qwertyquerty) [inline code/modules/power/supermatter/supermatter_cascade_escape.dm:42]: Remove

File: code/modules/power/supermatter/supermatter_cascade.dm [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	sleep(100)
+	SSshuttle.emergency.cancel(null)
+	sleep(50)
+	priority_announce("Attention, [station_name()]. A u̢nivér͜sal͝-͢wi͠de ͞EMP͟ ̀has bee̡n d͟etec̕t̀ed͝. ͘M͜any̴ ͘sys̸te̡ms ͞at ͞C͡e͞n̶tC̨o̶m a͏re ͢f́ai͜l͠iǹg or h̷a͟ve ͏failed̛ an̨d m̢a̧ņy pe͟r҉so͞n͘n̡e̵l ͢are ͏d҉e҉ad́.̡ W͞e h̸av͘e d̀e͘tect͏ed sign̴s̵ ͘óf ̷t̷ot͝a҉l̛ immi̴nen͢t r̵e̴a͢lity f͝a͞ilure. ̧However̶, ͜you͝ ̕a͏ŕe ́t҉he͞ on҉l͘y st̵a̴t̨i͟o͘n̸ ͞n̸ear̢ ͞a ̛b͝l͟u̡es̀p͝a͝c̢e̛ ͟r̵if͠t̡ to a ͜shiel̀d͘ed̨ ̨ùn͠i̕ve͢rs͜e͜. Goo̷d ͠lu̷ćk ̸mak͘in͠g͘ it t͡he̵re, ͞c̴rew̢.","*enfra̕l̸ ̸Com^and̡ ҉Em/r͟ge͘n͟c͢y C̴om%$ni̸c̨a͏t̡i-ǹs̨")
+	sleep(50)
+	for(var/mob/M in GLOB.player_list)
+		if(ishuman(M))
+			var/datum/antagonist/supermatter_survivalist/survivalist = new
+			M.mind.add_antag_datum(survivalist)
+	//SSshuttle.registerHostileEnvironment(GLOB.supermatter_cascade)
```
Comment (qwertyquerty) [inline code/modules/power/supermatter/supermatter_cascade.dm:54]: Remove comment

File: code/modules/power/supermatter/supermatter_cascade.dm [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+/obj/machinery/power/supermatter_crystal
+	var/cascade_allowed = TRUE
+	//var/cascade_time = 0
+/*
+/obj/machinery/power/supermatter_crystal/cascade
+	name = "energetic supermatter crystal"
+	desc = "A strangely translucent and iridescent crystal. It looks very energetic. It'd be unwise to let it delaminate."
+	cascade_allowed = TRUE
+	config_bullet_energy = 0.1 //No cheesing it with guns
+*/
```
Comment (qwertyquerty) [inline code/modules/power/supermatter/supermatter_cascade.dm:12]: Remove comments

**Comment by PowerfulBacon** [comment]
Literally an easy grief method for destroying the round that is completely unstoppable by admins and destroys the round for everyone no matter what, with no reason for it and by our rules will mean if it gets activated, someone will be banned no matter what. 
The only valid things for this is griefing since even if an antag does it they would probably get banned, and anyone can do this easilly.

**Comment by coolsurf6** [comment]
Ok ok @PowerfulBacon that was more of a joke. BUT it can still maybe be somewhat useful. Basically if you throw a telecrystal into it, it may increase power usage to some extreme amount, and might make for a fun project for engineers to make massive power stonks. But if it were to delam in that state, that wacky reality tearing effect happens. 

Maybe it will only be in that state for 5 minutes or so, or some other more specific item needs to be met. Or just have it do the more devestating reality tear if it is incredibly incredibly hot. It does make the SM blowing up a bit more of a concern than just "eh, don't go near engineering and get in lockers, yall be fine".

**Comment by PowerfulBacon** [comment]
> 
> 
> Ok ok @PowerfulBacon that was more of a joke. BUT it can still maybe be somewhat useful. Basically if you throw a telecrystal into it, it may increase power usage to some extreme amount, and might make for a fun project for engineers to make massive power stonks. But if it were to delam in that state, that wacky reality tearing effect happens.
> 
> Maybe it will only be in that state for 5 minutes or so, or some other more specific item needs to be met. Or just have it do the more devestating reality tear if it is incredibly incredibly hot. It does make the SM blowing up a bit more of a concern than just "eh, don't go near engineering and get in lockers, yall be fine".

I was talking about this effect in general being only useful for griefing, but if there was a positive to doing something like that instead of being only danger, that would be more acceptable.

**Comment by coolsurf6** [comment]
The positive is just that you are gaining a literal fuck ton of power. But the down side is, seriously don't delam it. It would take an engineer to have an already set up and well made SM for it to even be possible as the threshold is massive.

**Comment by coolsurf6** [comment]
A minge throwing themselves into the SM isn't going to make that activate.

**Comment by PowerfulBacon** [comment]
My problem is, its a relatively easy thing for a griefer to do if they want to fuck over everyone's round instantly.
If an antag were to do this, they would probably be banned unless they had martyr and if a non-antag did this they would probably get banned.

**Comment by KingTerrytheTerrible** [comment]
> 
> 
> My problem is, its a relatively easy thing for a griefer to do if they want to fuck over everyone's round instantly.
> If an antag were to do this, they would probably be banned unless they had martyr and if a non-antag did this they would probably get banned.

It's not that easy to do. An antag and a griefer both have to increase the energy levels past 11000KeV and keep it there for around 7 minutes, I'd estimate. You'd have security AND engineering up your ass trying to stop it.

**Comment by KingTerrytheTerrible** [comment]
It also gives engineering an incentive to give a shit about the engine failing; especially if there's a traitor engi.

**Pull request closed without merging**